### ISSUE TITLE
Do not add default `node`, `browser`, `require` or `import` conditions

### ISSUE THREAD

**Issue by SimenB** [lukeed/resolve.exports#12]
I'm looking into using this module to implement support for `exports` in Jest.

Seems unfortunate that Node's defaults should be forced on everybody using `exports` field. For a module that isn't supported to run in Node, it should just fail to resolve or something (or falling back to `default`) not pick up `node`.

There's already a switch in this library for `browser` condition so it's not consistently enforced (at least from my understanding `browser` is just a thing Node acknowledges exists, nowhere in the spec does it say its presence removes the `node` condition) - why not just fully let the caller provide their own conditions (barring possibly `default` - that seems reasonable to always add)? I.e. remove the two `allows.add` calls (and maybe the options leaving only `conditions`). Or possibly add `node` if no `conditions` is passed, but otherwise add nothing.

As a datapoint, TypeScript could not have used `resolve.exports` to implement support for `types` condition with their official example: https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#packagejson-exports-imports-and-self-referencing since this module would have resolved `import` first.

I think `resolve.exports` should be implementation/host independent (maybe a `resolve.exports-node` or just a `import { node } from 'resolve.exports'` makes sense, but the base export shouldn't put these limitations there).

Tl;dr: removing https://github.com/lukeed/resolve.exports/blob/c6814c43cb27ee6092f649be3fdf6c1fb55c51e4/src/index.js#L71-L72 would allow this module to be used by any runtime/host to implement `exports` support without getting Node semantics/conditions when the module might not even run in Node or a browser.

---

This is mostly a copy paste of https://github.com/yarnpkg/berry/pull/2431#discussion_r720221992.

**Comment by ljharb** [comment]
Indeed, `default` must not always exist. node's docs are that way to try to obscure all the things that the authors don't want you using, but all of my packages use `default`.

For `require`, i believe the first condition it finds among node/require/default will be used; for `import`, import/default?

**Comment by ljharb** [comment]
@lukeed that's not the case in older versions of node, which don't understand `require` or `import` at all. Check in node v13.2-v13.6, for example.

**Comment by ljharb** [comment]
@lukeed i absolutely agree they should all get patch fixes, but node has declined to do so. whether they "should" be supported is each author's choice; every package i maintain will continue to work with every possible node minor for as long as I maintain them :-)

It sounds like as long as a `browser` condition goes first, say, then a tool looking for the `browser` entrypoint will get it?

**Comment by lukeed** [comment]
Node 13 is a dead version – and **literally no one** has any reason to continue using it. I do not say this as live-on-the-cutting-edge type of person. I generally hold onto support for LTS versions of Node longer than the average developer, but odd-number versions are **purposefully** experimental. Node itself does not recommend that any one use odd versions in production or beyond the next even-version's release. I'm sure you're familiar with this, so I'm not sure why you are trying to advocate for this idea? Even Node 14 [enters "maintenance"](https://nodejs.org/en/about/releases/) in 7 days. Anyone on Node 9/11/13 deserves a bad experience at this point.

> It sounds like as long as a browser condition goes first, say, then a tool looking for the browser entrypoint will get it?

Yes, I've laid out a number of examples explaining this. If you only have time to skim one, I'd recommend you look at the output for the `resolve(example, ['browser', 'development', 'import']);` case above.

The `exports` key order defines the priority.
The resolving conditions defines keys the consumer is willing to accept, if defined.

**Comment by lukeed** [comment]
>  it just places the burden onto package authors whose packages are used in these envs to always put require/import last 

As mentioned, this is a two-way street. Package authors are required to have some awareness of their exports definition.

But you're right, more often than not you'll be defining the `import` and `require` conditions last in the object and defining environment-specific conditions first. This is why key-order matters and also why nested conditions are allowed:

```
exports: {
  env1: {
    import: ..
    require: ...
  },
  env2: {
    // ...
  },
  import: ...
  require: ...
}
```

Here's probably my most complex public example; [`httpie@next`](https://github.com/lukeed/httpie/blob/next/package.json#L13). When importing the root module, a resolver aware of the "worker" condition will always get that code, else browser (if wanted), else it's Node with import-type used as differentiator. I could have had nested import/require conditions within "worker" and "browser", but as an author I'm aware of the fact that users will be using bundlers in worker/browser environments, meaning I only want to offer them esm anyway.

> And even if those conditions were put last, if my runtime doesn't support require or import (and isn't browser or node), but defines its own condition with a custom module system and I then load a module who inly defines those, I'd expect it to fall back to attempt to load the conditions I define (plus default probably). If I were to use this module to do resolving of exports, I would get one of those conditions I don't support back

You'd only ever get a result/match for something you said you supported and/or wanted. Aka, in the httpie example, no one will ever get the "worker" file unless it was explicitly allowed. That's the purpose/safety of custom conditions -- author offers standard X conditions but also offers custom, optional Y conditions (worker, react-native, electron... *cough* types *cough* lol) that relevant users can opt into since theyre the target.

As an author, if you define this:

```js
"exports": {
  ".": {
    "foobar": "./foobar.mjs"
  }
}
``` 

No one will ever see that file unless they've enabled the "foobar" condition (both with this module and with Node natively). Instead, the root import will resolve to nothing/undefined by default, which surfaces as `undefined` in this module and throws an Error in Node.

However, doing this:

```js
"exports": {
  ".": {
    "foobar": "./foobar.mjs",
    "default": "./fallback.js"
  }
}
``` 
...will resolve to the "fallback.js" file by default for this module and Node, avoiding the undefined/Error.

---

If you **really** need me to (99.9999% you don't though), I can add an `unsafe` option that clears the `allows` conditions set so that there are no defaults.

I'd want to keep the name ugly/scary because it's purposefully breaking out of defined behavior. Again, this really should not be needed by anybody who's actually willing to use/work with the exports field.

**Comment by lukeed** [comment]
Released in 1.1.0~! Hopefully added enough warning labels for the general use case, but I guess time will tell 😆 Guess this means that TS _could_ have used this now directly, although I'm sure they already have their recursive `if ("types" in thing)` checker in place.

**Comment by SimenB** [comment]
Thank you!
